File: Application/src/main/java/com/shareyourproxy/app/adapter/EditGroupChannelAdapter.java
package com.shareyourproxy.app.adapter;

import android.content.Context;
import android.support.design.widget.TextInputLayout;
import android.support.v7.util.SortedList;
import android.support.v7.util.SortedList.Callback;
import android.text.Editable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.style.TextAppearanceSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.Switch;
import android.widget.TextView;

import com.shareyourproxy.R;
import com.shareyourproxy.api.domain.model.Channel;
import com.shareyourproxy.api.domain.model.ChannelToggle;
import com.shareyourproxy.api.domain.model.ChannelType;
import com.shareyourproxy.api.rx.RxBusDriver;
import com.shareyourproxy.api.rx.RxGroupChannelSync;
import com.shareyourproxy.api.rx.event.ViewGroupContactsEvent;
import com.shareyourproxy.app.EditGroupChannelsActivity.GroupEditType;
import com.shareyourproxy.util.ObjectUtils;
import com.shareyourproxy.widget.DismissibleNotificationCard;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Locale;
import java.util.Map;

import butterknife.Bind;
import butterknife.BindString;
import timber.log.Timber;

import static com.shareyourproxy.app.EditGroupChannelsActivity.GroupEditType.EDIT_GROUP;
import static com.shareyourproxy.app.adapter.BaseViewHolder.ItemClickListener;
import static com.shareyourproxy.widget.DismissibleNotificationCard.NotificationCard.PUBLIC_GROUPS;

public class EditGroupChannelAdapter extends BaseRecyclerViewAdapter {
    public static final int TYPE_LIST_ITEM = 1;
    public static final int TYPE_LIST_HEADER = 2;
    public static final int TYPE_LIST_DELETE_FOOTER = 3;
    private final GroupEditType _groupEditType;
    private final RxBusDriver _rxBus = RxBusDriver.INSTANCE;
    private ItemClickListener _clickListener;
    private Callback<ChannelToggle> _sortedListCallback;
    private SortedList<ChannelToggle> _channels;
    private String _groupLabel;
    private TextWatcher _textWatcher = getTextWatcher();
    private HeaderViewHolder _groupLabelHeaderViewHolder;

    public EditGroupChannelAdapter(
        ItemClickListener listener, String groupLabel, HashMap<String, Channel> userChannels,
        HashSet<String> groupChannels, GroupEditType groupEditType) {
        _clickListener = listener;
        _groupLabel = groupLabel == null ? "" : groupLabel;
        _groupEditType = groupEditType;
        _channels = new SortedList<>(
            ChannelToggle.class, getSortedCallback(), userChannels.size());
        if (_groupEditType.equals(GroupEditType.PUBLIC_GROUP)) {
            updatePublicChannels(userChannels);
        } else {
            updateChannels(userChannels, groupChannels);
        }
    }

    public static EditGroupChannelAdapter newInstance(
        ItemClickListener listener, String groupLabel, HashMap<String, Channel> userChannels,
        HashSet<String> groupChannels, GroupEditType groupEditType) {
        return new EditGroupChannelAdapter(listener, groupLabel,
            userChannels, groupChannels, groupEditType);
    }

    private TextWatcher getTextWatcher() {
        return new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                _groupLabel = s.toString();
            }
        };
    }

    private void updateChannels(
        HashMap<String, Channel> userChannels, HashSet<String> groupChannels) {
        if (userChannels != null) {
            ArrayList<ChannelToggle> channelToggles = new ArrayList<>();
            for (Map.Entry<String, Channel> userChannel : userChannels.entrySet()) {
                channelToggles.add(
                    new ChannelToggle(userChannel.getValue(),
                        channelInGroup(userChannel.getValue(), groupChannels)));
            }
            _channels.beginBatchedUpdates();
            for (ChannelToggle channel : channelToggles) {
                _channels.add(channel);
            }
            _channels.endBatchedUpdates();
        }
    }

    private void updatePublicChannels(HashMap<String, Channel> userChannels) {
        if (userChannels != null) {
            ArrayList<ChannelToggle> channelToggles = new ArrayList<>();
            for (Map.Entry<String, Channel> userChannel : userChannels.entrySet()) {
                Channel channel = userChannel.getValue();
                channelToggles.add(
                    new ChannelToggle(channel, channel.isPublic()));
            }
            _channels.beginBatchedUpdates();
            for (ChannelToggle channel : channelToggles) {
                _channels.add(channel);
            }
            _channels.endBatchedUpdates();
        }
    }

    private boolean channelInGroup(Channel userChannel, HashSet<String> groupChannels) {
        if (groupChannels != null) {
            for (String groupChannel : groupChannels) {
                if (groupChannel.equals(userChannel.id())) {
                    return true;
                }
            }
        }
        return false;
    }

    private Callback<ChannelToggle> getSortedCallback() {
        if (_sortedListCallback == null) {
            _sortedListCallback = new Callback<ChannelToggle>() {

                @Override
                public int compare(ChannelToggle item1, ChannelToggle item2) {
                    int weight1 = item1.getChannel().channelType().getWeight();
                    int weight2 = item2.getChannel().channelType().getWeight();
                    int compareFirst = ObjectUtils.compare(weight1, weight2);
                    if (compareFirst == 0) {
                        return item1.getChannel().label()
                            .compareToIgnoreCase(item2.getChannel().label());
                    } else {
                        return compareFirst;
                    }
                }

                @Override
                public void onInserted(int position, int count) {
                    notifyItemRangeInserted(position, count);
                }

                @Override
                public void onRemoved(int position, int count) {
                    notifyItemRangeRemoved(position, count);
                }

                @Override
                public void onMoved(int fromPosition, int toPosition) {
                    notifyItemMoved(fromPosition, toPosition);
                }

                @Override
                public void onChanged(int position, int count) {
                    notifyItemRangeChanged(position, count);
                }

                @Override
                public boolean areContentsTheSame(
                    ChannelToggle item1, ChannelToggle item2) {
                    return item1.getChannel().id().equals(item2.getChannel().id());
                }

                @Override
                public boolean areItemsTheSame(ChannelToggle item1, ChannelToggle item2) {
                    return item1.getChannel().id().equals(item2.getChannel().id());
                }
            };
        }
        return _sortedListCallback;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return TYPE_LIST_HEADER;
        } else if (_groupEditType.equals(EDIT_GROUP) && position == getListLength()) {
            return TYPE_LIST_DELETE_FOOTER;
        } else {
            return TYPE_LIST_ITEM;
        }
    }

    @Override
    public int getItemCount() {
        return getListSize();
    }

    /**
     * Get the list items length
     *
     * @return list length
     */
    private int getListLength() {
        return getExtraItemsCount() - 1;
    }

    /**
     * Get the list items size
     *
     * @return list size
     */
    private int getListSize() {
        return getExtraItemsCount();
    }

    private int getExtraItemsCount() {
        int count = 0;
        if (_groupEditType.equals(EDIT_GROUP)) {
            // show the header EditTextView and footer delete Button
            count = 2;
        } else if (_groupEditType.equals(GroupEditType.ADD_GROUP) ||
            _groupEditType.equals(GroupEditType.PUBLIC_GROUP)) {
            // just show the header EditTextView
            count = 1;
        }
        return _channels.size() + count;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view;
        if (viewType == TYPE_LIST_HEADER) {
            if (_groupEditType.equals(GroupEditType.PUBLIC_GROUP)) {
                view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.adapter_dismissible_notification, parent, false);
                return PublicHeaderViewHolder.newInstance(view, _clickListener);
            } else {
                view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.adapter_edit_group_channel_header, parent, false);
                return HeaderViewHolder.newInstance(view, _clickListener);
            }
        } else if (viewType == TYPE_LIST_DELETE_FOOTER) {
            view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.adapter_edit_group_channel_footer, parent, false);
            return FooterViewHolder.newInstance(view, _clickListener);
        } else {
            view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.adapter_edit_group_channel_item, parent, false);
            return ItemViewHolder.newInstance(view, _clickListener);
        }

    }

    @Override
    public void onBindViewHolder(BaseViewHolder holder, int position) {
        if (holder.getItemViewType() == TYPE_LIST_HEADER) {
            if (_groupEditType.equals(GroupEditType.PUBLIC_GROUP)) {
                bindPublicHeaderViewData((PublicHeaderViewHolder) holder);
            } else {
                bindHeaderViewData((HeaderViewHolder) holder);
            }
        } else if (holder.getItemViewType() == TYPE_LIST_ITEM &&
            (!_groupEditType.equals(GroupEditType.EDIT_GROUP) ||
                (position != getListLength()))) {
            bindItemViewData((ItemViewHolder) holder, getItemData(position - 1));
        }
    }

    public String getGroupLabel() {
        return _groupLabel;
    }

    public void bindPublicHeaderViewData(PublicHeaderViewHolder holder) {
        holder.notificationCard.createNotificationCard(this, holder, PUBLIC_GROUPS, false, false);
    }

    private void bindHeaderViewData(HeaderViewHolder holder) {
        Context context = holder.view.getContext();
        _groupLabelHeaderViewHolder = holder;
        holder.editText.setText(_groupLabel);
        holder.editText.addTextChangedListener(_textWatcher);
        int end = holder.title.length();
        SpannableStringBuilder sb = new SpannableStringBuilder(holder.title)
            .append("\n").append(holder.button.toUpperCase(Locale.US));
        TextAppearanceSpan titleSpan = new TextAppearanceSpan(context, R.style
            .Proxy_TextAppearance_Body_Disabled);
        TextAppearanceSpan buttonSpan = new TextAppearanceSpan(context, R.style
            .Proxy_TextAppearance_Button_Blue);
        sb.setSpan(titleSpan, 0, end, Spanned.SPAN_INCLUSIVE_INCLUSIVE);
        sb.setSpan(buttonSpan, end + 1, sb.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
        holder.textViewContacts.setText(sb);
        holder.textViewContacts.setOnClickListener(getContactsListener());
    }

    private View.OnClickListener getContactsListener() {
        return new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                _rxBus.post(new ViewGroupContactsEvent());
            }
        };
    }

    public ChannelToggle getItemData(int position) {
        if (position < _channels.size()) {
            return _channels.get(position);
        } else {
            return null;
        }
    }

    public void promptGroupLabelError(Context context) {
        _groupLabelHeaderViewHolder.
            textInputLayout.setError(context.getString(R.string.required));
        _groupLabelHeaderViewHolder.textInputLayout.setErrorEnabled(true);
    }

    private void bindItemViewData(ItemViewHolder holder, ChannelToggle editChannel) {
        Context context = holder.view.getContext();
        ChannelType channelType = editChannel.getChannel().channelType();
        Channel channel = editChannel.getChannel();
        String channelTypeString = editChannel.getChannel().channelType().getLabel();
        String label = editChannel.getChannel().label();
        String address = editChannel.getChannel().actionAddress();
        SpannableStringBuilder sb = getChannelSpannableStringBuilder(
            context, channelTypeString, label, address);

        holder.itemImage.setImageDrawable(
            getChannelIconDrawable(context, channel, getChannelBackgroundColor(context,
                channelType)));

        View.OnClickListener clickListener = switchListener(holder);
        CompoundButton.OnCheckedChangeListener checkedListener = checkedListener(holder);

        holder.itemLabel.setText(sb);
        holder.container.setOnClickListener(clickListener);
        holder.itemSwitch.setOnCheckedChangeListener(checkedListener);
        holder.itemSwitch.setChecked(editChannel.inGroup());
    }

    private View.OnClickListener switchListener(final ItemViewHolder viewHolder) {
        return new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int position = viewHolder.getItemPosition() - 1;
                Switch itemSwitch = viewHolder.itemSwitch;
                Timber.i("position: %1$s", position);

                boolean toggle = !itemSwitch.isChecked();
                itemSwitch.setChecked(toggle);
            }
        };
    }

    private CompoundButton.OnCheckedChangeListener checkedListener(
        final ItemViewHolder
            viewHolder) {
        return new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                int position = viewHolder.getItemPosition() - 1;
                ChannelToggle channelToggle = _channels.get(position);
                channelToggle.setInGroup(isChecked);
            }
        };
    }

    public HashSet<String> getSelectedChannels() {
        return RxGroupChannelSync.INSTANCE.getSelectedChannels(_channels);
    }

    public ArrayList<ChannelToggle> getToggledChannels() {
        ArrayList<ChannelToggle> channels = new ArrayList<>(_channels.size());
        for (int i = 0; i < _channels.size(); i++) {
            ChannelToggle newChannel = _channels.get(i);
            channels.add(newChannel);
        }
        return channels;
    }

    public static final class ItemViewHolder extends BaseViewHolder {
        @Bind(R.id.adapter_edit_group_list_item_container)
        RelativeLayout container;
        @Bind(R.id.adapter_edit_group_list_item_switch)
        Switch itemSwitch;
        @Bind(R.id.adapter_edit_group_list_item_image)
        ImageView itemImage;
        @Bind(R.id.adapter_edit_group_list_item_label)
        TextView itemLabel;

        private ItemViewHolder(View view, ItemClickListener itemClickListener) {
            super(view, itemClickListener);
        }

        public static ItemViewHolder newInstance(View view, ItemClickListener itemClickListener) {
            return new ItemViewHolder(view, itemClickListener);
        }

    }

    public static final class PublicHeaderViewHolder extends BaseViewHolder {
        @Bind(R.id.adapter_dismissible_notification_card)
        DismissibleNotificationCard notificationCard;

        private PublicHeaderViewHolder(View view, ItemClickListener itemClickListener) {
            super(view, itemClickListener);
        }

        public static PublicHeaderViewHolder newInstance(
            View view, ItemClickListener itemClickListener) {
            return new PublicHeaderViewHolder(view, itemClickListener);
        }

    }

    public static final class HeaderViewHolder extends BaseViewHolder {
        @Bind(R.id.adapter_group_edit_channel_header_edittext)
        EditText editText;
        @Bind(R.id.adapter_group_edit_channel_header_floatlabel)
        TextInputLayout textInputLayout;
        @Bind(R.id.adapter_group_edit_channel_header_contacts_button)
        TextView textViewContacts;
        @BindString(R.string.people_in_this_group)
        String title;
        @BindString(R.string.view_group_members)
        String button;

        private HeaderViewHolder(View view, ItemClickListener itemClickListener) {
            super(view, itemClickListener);
        }

        public static HeaderViewHolder newInstance(View view, ItemClickListener itemClickListener) {
            return new HeaderViewHolder(view, itemClickListener);
        }

    }

    public static final class FooterViewHolder extends BaseViewHolder {
        @Bind(R.id.adapter_group_edit_channel_footer_delete)
        Button deleteButton;

        private FooterViewHolder(View view, ItemClickListener itemClickListener) {
            super(view, itemClickListener);
        }

        public static FooterViewHolder newInstance(View view, ItemClickListener itemClickListener) {
            return new FooterViewHolder(view, itemClickListener);
        }

    }
}

File: Application/src/main/kotlin/com/shareyourproxy/app/adapter/EditGroupChannelAdapter.kt
package com.shareyourproxy.app.adapter

import android.content.Context
import android.support.design.widget.TextInputLayout
import android.support.v7.util.SortedList
import android.support.v7.util.SortedList.Callback
import android.text.Editable
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.TextWatcher
import android.text.style.TextAppearanceSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.shareyourproxy.R
import com.shareyourproxy.R.id.*
import com.shareyourproxy.R.string.required
import com.shareyourproxy.api.domain.model.Channel
import com.shareyourproxy.api.domain.model.ChannelToggle
import com.shareyourproxy.api.rx.RxBusRelay.post
import com.shareyourproxy.api.rx.RxGroupChannelSync.getSelectedChannels
import com.shareyourproxy.api.rx.event.ViewGroupContactsEvent
import com.shareyourproxy.app.EditGroupChannelsActivity.GroupEditType
import com.shareyourproxy.app.EditGroupChannelsActivity.GroupEditType.EDIT_GROUP
import com.shareyourproxy.app.adapter.BaseViewHolder.ItemClickListener
import com.shareyourproxy.util.ButterKnife.bindString
import com.shareyourproxy.util.ButterKnife.bindView
import com.shareyourproxy.widget.DismissibleNotificationCard
import com.shareyourproxy.widget.DismissibleNotificationCard.NotificationCard.PUBLIC_GROUPS
import timber.log.Timber
import java.util.*

class EditGroupChannelAdapter(private val recyclerView: BaseRecyclerView, private val clickListener: ItemClickListener, internal var groupLabel: String, userChannels: HashMap<String, Channel>, groupChannels: HashSet<String>, private val groupEditType: GroupEditType) : BaseRecyclerViewAdapter() {
    companion object{
        internal val TYPE_LIST_ITEM = 1
        internal val TYPE_LIST_HEADER = 2
        internal val TYPE_LIST_DELETE_FOOTER = 3
    }
    private val channels: SortedList<ChannelToggle> = SortedList(ChannelToggle::class.java, sortedCallback, userChannels.size)
    private val stringTitle: String by bindString(recyclerView.context, R.string.people_in_this_group)
    private val stringButton: String by bindString(recyclerView.context, R.string.view_group_members)
    private val contactsListener: View.OnClickListener = View.OnClickListener { post(ViewGroupContactsEvent()) }
    internal val selectedChannels: HashSet<String> get()= getSelectedChannels(channels)
    private var groupLabelHeaderViewHolder: HeaderViewHolder? = null

    init {
        if (groupEditType == GroupEditType.PUBLIC_GROUP) {
            updatePublicChannels(userChannels)
        } else {
            updateChannels(userChannels, groupChannels)
        }
    }

    private val textWatcher: TextWatcher get() = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

        }

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {

        }

        override fun afterTextChanged(s: Editable) {
            groupLabel = s.toString()
        }
    }

    /**
     * Get the list items length
     * @return list length
     */
    private val listLength: Int = extraItemsCount - 1

    private val extraItemsCount: Int get() {
        var count = 0
        if (groupEditType == EDIT_GROUP) {
            count = 2
        } else if (groupEditType == GroupEditType.ADD_GROUP || groupEditType == GroupEditType.PUBLIC_GROUP) {
            count = 1
        }
        return channels.size() + count
    }

    internal val toggledChannels: ArrayList<ChannelToggle> get() {
        val channels = ArrayList<ChannelToggle>(channels.size())
        for (i in 0..channels.size - 1) {
            val newChannel = channels[i]
            channels.add(newChannel)
        }
        return channels
    }

    private val sortedCallback: Callback<ChannelToggle> get() = object : Callback<ChannelToggle>() {
        override fun compare(item1: ChannelToggle, item2: ChannelToggle): Int {
            val weight1 = item1.channel.channelType.weight
            val weight2 = item2.channel.channelType.weight
            val compareFirst = compareValues(weight1, weight2)
            if (compareFirst == 0) {
                return item1.channel.label.compareTo(item2.channel.label, true)
            } else {
                return compareFirst
            }
        }

        override fun onInserted(position: Int, count: Int) {
            notifyItemRangeInserted(position, count)
        }

        override fun onRemoved(position: Int, count: Int) {
            notifyItemRangeRemoved(position, count)
        }

        override fun onMoved(fromPosition: Int, toPosition: Int) {
            notifyItemMoved(fromPosition, toPosition)
        }

        override fun onChanged(position: Int, count: Int) {
            notifyItemRangeChanged(position, count)
        }

        override fun areContentsTheSame(
                item1: ChannelToggle, item2: ChannelToggle): Boolean {
            return item1.channel.id.equals(item2.channel.id)
        }

        override fun areItemsTheSame(item1: ChannelToggle, item2: ChannelToggle): Boolean {
            return item1.channel.id.equals(item2.channel.id)
        }
    }

    override fun getItemViewType(position: Int): Int {
        if (position == 0) {
            return TYPE_LIST_HEADER
        } else if (groupEditType == EDIT_GROUP && position == listLength) {
            return TYPE_LIST_DELETE_FOOTER
        } else {
            return TYPE_LIST_ITEM
        }
    }

    override fun getItemCount(): Int {
        return extraItemsCount
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        val view: View
        if (viewType == TYPE_LIST_HEADER) {
            if (groupEditType == GroupEditType.PUBLIC_GROUP) {
                view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_dismissible_notification, parent, false)
                return PublicHeaderViewHolder(view, clickListener)
            } else {
                view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_edit_group_channel_header, parent, false)
                return HeaderViewHolder(view, clickListener)
            }
        } else if (viewType == TYPE_LIST_DELETE_FOOTER) {
            view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_edit_group_channel_footer, parent, false)
            return FooterViewHolder(view, clickListener)
        } else {
            view = LayoutInflater.from(parent.context).inflate(R.layout.adapter_edit_group_channel_item, parent, false)
            return ItemViewHolder(view, clickListener)
        }

    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        if (holder.itemViewType == TYPE_LIST_HEADER) {
            if (groupEditType == GroupEditType.PUBLIC_GROUP) {
                bindPublicHeaderViewData(holder as PublicHeaderViewHolder)
            } else {
                bindHeaderViewData(holder as HeaderViewHolder)
            }
        } else if (holder.itemViewType == TYPE_LIST_ITEM && (groupEditType != GroupEditType.EDIT_GROUP || (position != listLength))) {
            bindItemViewData(holder as ItemViewHolder, getItemData(position - 1))
        }
    }

    private fun updateChannels(userChannels: HashMap<String, Channel>?, groupChannels: HashSet<String>) {
        if (userChannels != null) {
            val channelToggles = ArrayList<ChannelToggle>()
            for (userChannel in userChannels.entries) {
                channelToggles.add(ChannelToggle(userChannel.value, channelInGroup(userChannel.value, groupChannels)))
            }
            channels.beginBatchedUpdates()
            for (channel in channelToggles) {
                channels.add(channel)
            }
            channels.endBatchedUpdates()
        }
    }

    private fun updatePublicChannels(userChannels: HashMap<String, Channel>?) {
        if (userChannels != null) {
            val channelToggles = ArrayList<ChannelToggle>()
            for (userChannel in userChannels.entries) {
                val channel = userChannel.value
                channelToggles.add(
                        ChannelToggle(channel, channel.isPublic))
            }
            channels.beginBatchedUpdates()
            for (channel in channelToggles) {
                channels.add(channel)
            }
            channels.endBatchedUpdates()
        }
    }

    private fun channelInGroup(userChannel: Channel, groupChannels: HashSet<String>?): Boolean {
        if (groupChannels != null) {
            for (groupChannel in groupChannels) {
                if (groupChannel == userChannel.id) {
                    return true
                }
            }
        }
        return false
    }

    private fun bindPublicHeaderViewData(holder: PublicHeaderViewHolder) {
        holder.notificationCard.createNotificationCard(this, holder, PUBLIC_GROUPS, false, false)
    }

    private fun bindHeaderViewData(holder: HeaderViewHolder) {
        val context = holder.view.context
        groupLabelHeaderViewHolder = holder
        holder.editText.setText(groupLabel)
        holder.editText.addTextChangedListener(textWatcher)
        val end = stringTitle.length
        val sb = SpannableStringBuilder(stringTitle).append("\n").append(stringButton.toUpperCase(Locale.US))
        val titleSpan = TextAppearanceSpan(context, R.style.Proxy_TextAppearance_Body_Disabled)
        val buttonSpan = TextAppearanceSpan(context, R.style.Proxy_TextAppearance_Button_Blue)
        sb.setSpan(titleSpan, 0, end, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
        sb.setSpan(buttonSpan, end + 1, sb.length, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
        holder.textViewContacts.text = sb
        holder.textViewContacts.setOnClickListener(contactsListener)
    }

    internal fun getItemData(position: Int): ChannelToggle {
        return channels.get(position)
    }

    internal fun promptGroupLabelError(context: Context) {
        groupLabelHeaderViewHolder!!.textInputLayout.error = context.getString(required)
        groupLabelHeaderViewHolder!!.textInputLayout.isErrorEnabled = true
    }

    private fun bindItemViewData(holder: ItemViewHolder, editChannel: ChannelToggle) {
        val context = holder.view.context
        val channelType = editChannel.channel.channelType
        val channel = editChannel.channel
        val channelTypeString = editChannel.channel.channelType.label
        val label = editChannel.channel.label
        val address = editChannel.channel.actionAddress
        val sb = getChannelSpannableStringBuilder(context, channelTypeString, label, address)
        val clickListener = switchListener(holder)
        val checkedListener = checkedListener(holder)

        holder.itemImage.setImageDrawable(getChannelDrawable(channel, channelType, context))
        holder.itemLabel.text = sb
        holder.container.setOnClickListener(clickListener)
        holder.itemSwitch.setOnCheckedChangeListener(checkedListener)
        holder.itemSwitch.isChecked = editChannel.inGroup
    }

    private fun switchListener(viewHolder: ItemViewHolder): View.OnClickListener {
        return View.OnClickListener {
            val position = viewHolder.itemPosition - 1
            val itemSwitch = viewHolder.itemSwitch
            Timber.i("position: $position")

            val toggle = !itemSwitch.isChecked
            itemSwitch.isChecked = toggle
        }
    }

    private fun checkedListener(viewHolder: ItemViewHolder): CompoundButton.OnCheckedChangeListener {
        return CompoundButton.OnCheckedChangeListener { buttonView, isChecked ->
            val position = viewHolder.itemPosition - 1
            val channelToggle = channels.get(position)
            channelToggle.inGroup = isChecked
        }
    }

    private final class ItemViewHolder(view: View, itemClickListener: ItemClickListener) : BaseViewHolder(view, itemClickListener) {
        val container: RelativeLayout by bindView(adapter_edit_group_list_item_container)
        val itemSwitch: Switch by bindView(adapter_edit_group_list_item_switch)
        val itemImage: ImageView by bindView(adapter_edit_group_list_item_image)
        val itemLabel: TextView by bindView(adapter_edit_group_list_item_label)
    }

    private final class PublicHeaderViewHolder(view: View, itemClickListener: ItemClickListener) : BaseViewHolder(view, itemClickListener) {
        val notificationCard: DismissibleNotificationCard by bindView(adapter_dismissible_notification_card)
    }

    private final class HeaderViewHolder(view: View, itemClickListener: ItemClickListener) : BaseViewHolder(view, itemClickListener) {
        val editText: EditText by bindView(adapter_group_edit_channel_header_edittext)
        val textInputLayout: TextInputLayout by bindView(adapter_group_edit_channel_header_floatlabel)
        val textViewContacts: TextView by bindView(adapter_group_edit_channel_header_contacts_button)
    }

    private final class FooterViewHolder(view: View, itemClickListener: ItemClickListener) : BaseViewHolder(view, itemClickListener) {
        val deleteButton: Button by bindView(adapter_group_edit_channel_footer_delete)
    }
}
